packages/core/contracts/insured-bridge/BridgePool.sol

```diff
@@ -335,28 +328,73 @@ contract BridgePool is Testable, BridgePoolInterface, ExpandedERC20, MultiCaller
         // of the LP fees, the relayer fees and the instant relay fee.
         uint256 feesTotal =
             _getAmountFromPct(
-                relay.realizedLpFeePct + depositData.slowRelayFeePct + depositData.instantRelayFeePct,
+                relayData.realizedLpFeePct + depositData.slowRelayFeePct + depositData.instantRelayFeePct,
                 depositData.amount
             );
 
         l1Token.safeTransferFrom(msg.sender, depositData.l1Recipient, depositData.amount - feesTotal);
 
-        emit RelaySpedUp(depositHash, msg.sender, relay.realizedLpFeePct);
+        emit RelaySpedUp(depositHash, msg.sender, relayData);
     }
 
     /**
      * @notice Reward relayers if a pending relay price request has a price available on the OptimisticOracle. Mark
      * the relay as complete.
      * @param depositData Unique set of L2 deposit data that caller is trying to settle a relay for.
+     * @param relayData Parameters of Relay that caller is attempting to settle. Must hash to the stored relay hash
+     * for this deposit.
+     * @param ancillaryData Relay price request's ancillary data. Used to check status of relay price request.
+     * @param request Relay price request struct. Used to check status of relay price request.
      */
-    function settleRelay(DepositData memory depositData) public nonReentrant() {
+    function settleRelay(
+        DepositData memory depositData,
+        RelayData memory relayData,
+        bytes memory ancillaryData,
+        SkinnyOptimisticOracleInterface.Request memory request
+    ) public nonReentrant() {
         bytes32 depositHash = _getDepositHash(depositData);
-        RelayData storage relay = relays[depositHash];
-
-        require(relays[depositHash].relayState == RelayState.PendingFinalization, "Settle iff price resolved True");
+        _validateRelayDataHash(depositHash, relayData);
+        require(relayData.relayState == RelayState.Pending, "Already settled");
+
+        // Optimistically try to settle relay. If this transaction settles the relay, then check the newly resolved
+        // price. Otherwise grab the already resolved price.
+        try
+            optimisticOracle.settle(
```

Comment:
Can we avoid the try catch by:
```
if (!request.settled) {
  (, in256 resolvedPrice) = optimisticOracle.settle(...);
  require(resolvedPrice == int256(1e18), "Settle relay iff price is True");
else {
  require(optimisticOracle.hasPrice(...) && request.resolvedPrice == int256(1e18));
}
```

Not sure if try-catch is cheaper, but I assume it isn't. Probably worth a quick test.

Reply:
This is about 500 gas cheaper for `settleRelay` on average. Question is, do you think checking the `request.settled` param opens the attack surface? Or are the assumptions the same about the link between `depositData`, `relayData`, `ancillaryData`, and `request`?

Reply:
I think the assumptions are the same because in both cases we're asking the OO to verify that those request params match a valid request, right?

Reply:
Yeah fair, done and commented. Also added a comment about pros/cons to trying to derive the `request` from just the contract state + function params